app/lib/search/index_ducene.dart

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io';
+import 'dart:math';
+
+import 'package:ducene/analysis.dart';
+import 'package:ducene/document.dart';
+import 'package:ducene/search.dart';
+import 'package:ducene/util.dart';
+import 'package:gcloud/service_scope.dart' as ss;
+
+import '../shared/configuration.dart';
+import '../shared/search_service.dart';
+
+/// The [PackageIndex] registered in the current service scope.
+PackageIndex get packageIndex => ss.lookup(#_ducene_package_index);
+
+/// Register a new [PackageIndex] in the current service scope.
+void registerPackageIndex(PackageIndex index) =>
+    ss.register(#_ducene_package_index, index);
+
+class DucenePackageIndex implements PackageIndex {
+  Directory _fsDir;
+  IndexHolder _index;
+  DateTime _lastUpdated;
+  final _MultiNgramAnalyzer _analyzer = new _MultiNgramAnalyzer(3);
+
+  Future _init() async {
+    if (_index != null) return;
+    IndexHolderDirectory indexDir;
+    if (envConfig.duceneDir != null) {
+      // Temporarily everything is in memory, because ducene is not using async
+      // IO, and that blocks appengine's health checks, killing the machine
+      // before it got any chance to setup.
+      // TODO: fix ducene to handle async FS properly
+      // _fsDir = new Directory(envConfig.duceneDir);
+      // indexDir = new FSIndexHolderDirectory(_fsDir);
+      indexDir = new RAMIndexHolderDirectory();
+    } else {
+      indexDir = new RAMIndexHolderDirectory();
+    }
+    _index = await DirectoryHolder.open(indexDir);
+  }
```

Comment:
There is a race, when two calls to `_init()` are in the middle of `DirectoryHolder.open()` only the last one will win and set the `_index`.

Consider just doing something like this:

```dart
class ... {

  Future readFuture;

  Future _init()  {
    if (readyFuture != null) return readyFuture;
    final Completer completer = new Completer();
    readyFuture = completer.future;

    new Future(() async {
      ...
      await ...;
      completer.complete();
    });
  }
}
```